src/WorkerChannel.ts

```diff
@@ -33,32 +29,32 @@ export class WorkerChannel {
         });
     }
 
-    /**
-     * Register a patching function to be run before User Function is executed.
-     * Hook should return a patched version of User Function.
-     */
-    public registerBeforeInvocationRequest(beforeCb: InvocationRequestBefore): void {
-        this._invocationRequestBefore.push(beforeCb);
-    }
-
-    /**
-     * Register a function to be run after User Function resolves.
-     */
-    public registerAfterInvocationRequest(afterCb: InvocationRequestAfter): void {
-        this._invocationRequestAfter.push(afterCb);
+    public registerHook(hookName: string, callback: HookCallback): Disposable {
+        const hooks = this.getHooks(hookName);
+        hooks.push(callback);
+        return new Disposable(() => {
+            const index = hooks.indexOf(callback);
+            if (index > -1) {
+                hooks.splice(index, 1);
+            }
+        });
     }
 
-    public runInvocationRequestBefore(context: Context, userFunction: Function): Function {
-        let wrappedFunction = userFunction;
-        for (const before of this._invocationRequestBefore) {
-            wrappedFunction = before(context, wrappedFunction);
+    public async executeHooks(hookName: string, context: {}): Promise<void> {
+        const callbacks = this.getHooks(hookName);
+        for (const callback of callbacks) {
+            await callback(context);
         }
-        return wrappedFunction;
     }
 
-    public runInvocationRequestAfter(context: Context) {
-        for (const after of this._invocationRequestAfter) {
-            after(context);
+    private getHooks(hookName: string): HookCallback[] {
+        switch (hookName) {
+            case 'preInvocation':
+                return this._preInvocationHooks;
+            case 'postInvocation':
+                return this._postInvocationHooks;
+            default:

```

Comment:
## this instance

For this particular instance, yes I'm using string which would not be great on its own. _However_, the TypeScript types for "@azure/functions-worker" _are_ using string literals which helps a lot with intellisense. Specifically, the types package has stuff like this:
```
export function registerHook(
    hookName: 'postInvocation',
    callback: (context: PostInvocationContext) => void | Promise<void>
): Disposable;
```
I also have many unit tests, so a random typo here is unlikely to cause a regression in the future.

The reason I chose `string` here is because users _can_ pass any string at runtime. There is no guarantee it'll just be the hook names I currently have listed.

Reply:
So is there a benefit to having this line?

```TS
export function registerHook(hookName: string, callback: HookCallback): Disposable;
```

At least when using TS, leaving that part out would flag an error right away when using an unidentified hook, rather than on runtime.

Reply:
I can't think of a benefit if the TypeScript types and worker are always exactly in sync. _However_, I think the team's past history with typescript types shows how difficult it is to keep them in sync 😝 With a generic function like this, users can pass in a hookName that is supported by the worker even if the types package they're on doesn't have it yet.

I would describe this line as the "safe, flexible, easily backward compatible" option whereas the other lines are much more strict. There's benefits to both, but for the worker package I think it makes more sense to leans towards the former

Reply:
Makes sense, thanks!